### DISCUSSION TITLE
Use devices' drivers behind I2C mux (TCA9548A)

### DISCUSSION THREAD

**Discussion by mateela** [zephyrproject-rtos/zephyr#57027] [Q&A] [answered]
Hello, 

I want to use the TCA9548A I2C mux on my board, and mux several devices which I prepared my own drivers for. These drivers do the regular I2C calls RDs and WRs using the devices I2C slave address. 
If I understand correctly the driver I saw on the zephyr project for the TCA9548A mux device, it contains a small API which only allows to address the devices behind it by the transfer function. This function receives the pointer to the channel corresponds to the relevant device, and the requested message, and does 2 operations:
1. Enables the relevant channel of required device behind the mux.
2. Transfer the required message.
Example use can be found here: [https://github.com/zephyrproject-rtos/zephyr/issues/49046](url)

My issue here is that I don't want to generate the transaction to the endpoint devices through the mux driver, but through my drivers. My question is why the mux driver API does not contain the ability to simply change mux channel without sending message to the device (Just step 1 from the two above)? I want to address my devices using the drivers I made without them to be aware if the device is being accessed by mux with other devices or not. I understand that this means that I'll have to be responsible for ensuring the right mux channel before sending transaction toward it, but it worth the generalization of the drivers.

Thanks for the help,
Matan.

**Comment by GeorgeCGV** [comment, marked as answer]
Hello,

Based on #39589 the idea is to abstract mux usage from the underlying drivers.
I assume it shall work with custom drivers as long as they are plain I<sup>2</sup>C device drivers.

```
       mux: tca9546a@77 {
          compatible = "ti,tca9546a";
          reg = <0x77>;
          status = "okay";
          label = "i2c_mux";
          #address-cells = <1>;
          #size-cells = <0>;
          reset-gpios = <&gpio5 3 GPIO_ACTIVE_LOW>;
          mux_i2c@0 {
              compatible: "ti,tca9546a-channel"
              reg = <0>;
              #address-cells = <1>;
              #size-cells = <0>;
              temp_sens_0: tmp116@49 {
                  compatible = "ti,tmp116";
                  reg = <0x49>;
              };
          };
          mux_i2c@1 {
            compatible: "ti,tca9546a-channel"
              reg = <1>;
              #address-cells = <1>;
              #size-cells = <0>;
              temp_sens_1: tmp116@49 {
                  compatible = "ti,tmp116";
                  reg = <0x49>;
              };
          };
      };
```

That way devices don't care if they are operated with or without the mux. The abstraction of the device and mux looks reasonable for most use cases.

If there is a need and a use case to have a driver that controls the mux and underlying devices, then PRs are always welcomed.

**Reply by mateela** [reply]
Hi, thanks a lot for the response.

I still don't understand how in that way devices don't care if they are operated with or without the mux. For example, in the arrangement above - let's say the last operated channel in the mux is channel number 0 and I want to use the TMP116 in channel number 1. Specifically I want to do sample fetch from the sensor API - do I just call this function and the mux trapes this call and changes the channel automatically? or do I need to do it "manually" before the call to sample_fetch? and if I need to change the channel - how do I do it?

BR, 
Matan.

**Comment by ofirshe** [comment]
Hey @GeorgeCGV,

Based on the provided DTS node example, can we directly access the sensor API to examine the TMP116 device? Currently, it appears that our only option is to utilize the i2c_reg_read_byte/i2c_reg_write_byte functions and pass the desired channel as the device pointer (as mentioned in https://github.com/zephyrproject-rtos/zephyr/issues/49046 ).

Best regards

**Comment by ofirshe** [comment]
Hey @GeorgeCGV, 

Would you be able to provide some assistance with this?

**Reply by GeorgeCGV** [reply]
Hello, I don't have a device with that mux myself. But based on the code of the driver. The goal is to grab tmp device from DT and use it as if there is no mux.

Driver defines usable I2C device for each channel:

https://github.com/zephyrproject-rtos/zephyr/blob/68d1d7d1b4754a9cabde830dc378eebe28aed3a9/drivers/i2c/i2c_tca954x.c#L154-L171

each device handles mux channel selection on transfer:

https://github.com/zephyrproject-rtos/zephyr/blob/68d1d7d1b4754a9cabde830dc378eebe28aed3a9/drivers/i2c/i2c_tca954x.c#L91-L96

Have you tried to get defined temp sensor like `const struct device *sens0 = DEVICE_DT_GET(DT_NODELABEL(temp_sens_0));`?

**Reply by ofirshe** [reply]
Hey, I've tried and it works well as expected. 

Thanks
